File: Game/Level/SeededRNG.swift
import Foundation

struct SeededRNG {
    private var state: UInt64

    init(seed: UInt64) {
        state = seed == 0 ? 1 : seed
    }

    mutating func nextUInt64() -> UInt64 {
        state ^= state << 13
        state ^= state >> 7
        state ^= state << 17
        return state
    }

    mutating func next(bound: Int) -> Int {
        guard bound > 0 else { return 0 }
        return Int(nextUInt64() % UInt64(bound))
    }
}

File: android/app/src/main/java/com/uzls/four/level/SeededRNG.kt
package com.uzls.four.level

/**
 * XorShift64 PRNG — bit-identical to the Swift implementation.
 * Uses ULong for unsigned 64-bit arithmetic with the same shifts: shl 13, shr 7, shl 17.
 */
class SeededRNG(seed: ULong) {
    private var state: ULong = if (seed == 0UL) 1UL else seed

    fun nextUInt64(): ULong {
        state = state xor (state shl 13)
        state = state xor (state shr 7)
        state = state xor (state shl 17)
        return state
    }

    fun next(bound: Int): Int {
        if (bound <= 0) return 0
        return (nextUInt64() % bound.toULong()).toInt()
    }
}
